In [2]:
import pandas as pd

In [41]:
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, GridSearchCV, KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score, f1_score, cohen_kappa_score
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm

In [4]:
df = pd.read_csv('nba_logreg2.csv', sep = ";")
print(df.head(5))

              Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0  

[5 rows x 21 columns]


In [5]:
print(df.describe())

                GP          MIN          PTS          FGM          FGA  \
count  1329.000000  1329.000000  1329.000000  1329.000000  1329.000000   
mean     60.402558    17.647931     6.820166     2.634763     5.902483   
std      17.481092     8.327441     4.365671     1.686850     3.598820   
min      11.000000     3.100000     0.700000     0.300000     0.800000   
25%      47.000000    10.800000     3.700000     1.400000     3.300000   
50%      63.000000    16.100000     5.600000     2.100000     4.800000   
75%      77.000000    22.900000     8.800000     3.400000     7.500000   
max      82.000000    40.900000    28.200000    10.200000    19.800000   

               FG%      3P Made          3PA          3P%          FTM  \
count  1329.000000  1329.000000  1329.000000  1329.000000  1329.000000   
mean     44.117532     0.249661     0.785628    19.308126     1.302784   
std       6.119492     0.384608     1.063856    16.022916     0.989341   
min      23.800000     0.000000     0

In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1329 entries, 0 to 1328
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         1329 non-null   object 
 1   GP           1329 non-null   int64  
 2   MIN          1329 non-null   float64
 3   PTS          1329 non-null   float64
 4   FGM          1329 non-null   float64
 5   FGA          1329 non-null   float64
 6   FG%          1329 non-null   float64
 7   3P Made      1329 non-null   float64
 8   3PA          1329 non-null   float64
 9   3P%          1329 non-null   float64
 10  FTM          1329 non-null   float64
 11  FTA          1329 non-null   float64
 12  FT%          1329 non-null   float64
 13  OREB         1329 non-null   float64
 14  DREB         1329 non-null   float64
 15  REB          1329 non-null   float64
 16  AST          1329 non-null   float64
 17  STL          1329 non-null   float64
 18  BLK          1329 non-null   float64
 19  TOV   

## Parte 2

In [7]:
data = df[['MIN', 'PTS', 'FGM', 'FGA', 'FG%', '3P Made', '3PA','3P%', 'FTM', 'FTA', 'FT%', 'OREB', 'DREB', 'REB', 'AST', 'STL', 'BLK','TOV']]

In [10]:
#@title Estandarizar dataset

#Instanciar y aplicar fit_transfor a dataset
data_std =  StandardScaler().fit_transform(data)

# Generar dataframe desde los datos estandarizados, agregando nombre de columnas.
df_data_std = pd.DataFrame(data_std, columns=data.columns)

In [11]:
df_data_std.iloc[:,:-1].apply(lambda x: np.mean(x).round(2))

MIN        0.0
PTS        0.0
FGM       -0.0
FGA       -0.0
FG%        0.0
3P Made    0.0
3PA        0.0
3P%        0.0
FTM       -0.0
FTA        0.0
FT%        0.0
OREB       0.0
DREB      -0.0
REB       -0.0
AST       -0.0
STL        0.0
BLK       -0.0
dtype: float64

In [12]:
df_data_std.iloc[:,:-1].apply(lambda x: np.var(x))

MIN        1.0
PTS        1.0
FGM        1.0
FGA        1.0
FG%        1.0
3P Made    1.0
3PA        1.0
3P%        1.0
FTM        1.0
FTA        1.0
FT%        1.0
OREB       1.0
DREB       1.0
REB        1.0
AST        1.0
STL        1.0
BLK        1.0
dtype: float64

In [13]:
#@title Obtener matriz de correlación desde nuestr DataFrame Standarizado
matriz_corr = df_data_std.corr(method="pearson")
matriz_corr

MIN       PTS       FGM       FGA       FG%   3P Made       3PA  \
MIN      1.000000  0.911746  0.902864  0.910194  0.208341  0.389474  0.402814   
PTS      0.911746  1.000000  0.990874  0.979676  0.261587  0.345289  0.355198   
FGM      0.902864  0.990874  1.000000  0.980117  0.297113  0.287956  0.297898   
FGA      0.910194  0.979676  0.980117  1.000000  0.135815  0.388860  0.412053   
FG%      0.208341  0.261587  0.297113  0.135815  1.000000 -0.291573 -0.347680   
3P Made  0.389474  0.345289  0.287956  0.388860 -0.291573  1.000000  0.982579   
3PA      0.402814  0.355198  0.297898  0.412053 -0.347680  0.982579  1.000000   
3P%      0.165997  0.151072  0.119493  0.197160 -0.330690  0.589855  0.582337   
FTM      0.791089  0.896245  0.848146  0.826244  0.253381  0.155717  0.170463   
FTA      0.779813  0.880826  0.840667  0.805507  0.306360  0.093059  0.105784   
FT%      0.236160  0.254404  0.219423  0.264405 -0.150996  0.312256  0.320901   
OREB     0.573655  0.576421  0.597468  0.505547  0.512346 -0.218210 -0.231050   
DREB     0.747065  0.696560  0.705191  0.642752  0.411103  0.018688  0.013611   
REB      0.710853  0.679002  0.692678  0.616497  0.466126 -0.070865 -0.079111   
AST      0.629147  0.551561  0.532261  0.589038 -0.103546  0.374314  0.408066   
STL      0.757050  0.674937  0.662457  0.689771  0.061130  0.305146  0.336757   
BLK      0.401011  0.390496  0.400744  0.325503  0.390786 -0.155955 -0.169245   
TOV      0.826431  0.850272  0.834343  0.845952  0.126429  0.257015  0.282496   

              3P%       FTM       FTA       FT%      OREB      DREB       REB  \
MIN      0.165997  0.791089  0.779813  0.236160  0.573655  0.747065  0.710853   
PTS      0.151072  0.896245  0.880826  0.254404  0.576421  0.696560  0.679002   
FGM      0.119493  0.848146  0.840667  0.219423  0.597468  0.705191  0.692678   
FGA      0.197160  0.826244  0.805507  0.264405  0.505547  0.642752  0.616497   
FG%     -0.330690  0.253381  0.306360 -0.150996  0.512346  0.411103  0.466126   
3P Made  0.589855  0.155717  0.093059  0.312256 -0.218210  0.018688 -0.070865   
3PA      0.582337  0.170463  0.105784  0.320901 -0.231050  0.013611 -0.079111   
3P%      1.000000  0.030320 -0.032333  0.326372 -0.288759 -0.122949 -0.191071   
FTM      0.030320  1.000000  0.980626  0.253125  0.586325  0.657652  0.657194   
FTA     -0.032333  0.980626  1.000000  0.110846  0.655408  0.704195  0.714267   
FT%      0.326372  0.253125  0.110846  1.000000 -0.147512 -0.021901 -0.070799   
OREB    -0.288759  0.586325  0.655408 -0.147512  1.000000  0.838904  0.932874   
DREB    -0.122949  0.657652  0.704195 -0.021901  0.838904  1.000000  0.978199   
REB     -0.191071  0.657194  0.714267 -0.070799  0.932874  0.978199  1.000000   
AST      0.262120  0.474403  0.427141  0.292877 -0.010736  0.189380  0.121302   
STL      0.194329  0.599534  0.579737  0.202791  0.287982  0.413893  0.382942   
BLK     -0.242274  0.412565  0.473501 -0.160008  0.649201  0.687627  0.699615   
TOV      0.108277  0.804762  0.798856  0.195242  0.422708  0.572049  0.538216   

              AST       STL       BLK       TOV  
MIN      0.629147  0.757050  0.401011  0.826431  
PTS      0.551561  0.674937  0.390496  0.850272  
FGM      0.532261  0.662457  0.400744  0.834343  
FGA      0.589038  0.689771  0.325503  0.845952  
FG%     -0.103546  0.061130  0.390786  0.126429  
3P Made  0.374314  0.305146 -0.155955  0.257015  
3PA      0.408066  0.336757 -0.169245  0.282496  
3P%      0.262120  0.194329 -0.242274  0.108277  
FTM      0.474403  0.599534  0.412565  0.804762  
FTA      0.427141  0.579737  0.473501  0.798856  
FT%      0.292877  0.202791 -0.160008  0.195242  
OREB    -0.010736  0.287982  0.649201  0.422708  
DREB     0.189380  0.413893  0.687627  0.572049  
REB      0.121302  0.382942  0.699615  0.538216  
AST      1.000000  0.751109 -0.083851  0.747442  
STL      0.751109  1.000000  0.136144  0.742301  
BLK     -0.083851  0.136144  1.000000  0.284679  
TOV      0.747442  0.742301  0.284

En la prueba de Barlet, se obtiene un estadístico Chi2 de 53239.64 con 153 grados de libertad, dejando un p-value muy pequeño convergiendo a 0. Dicho esto, se puede rechazar la hipótesis nula a un 1% de significancia, lo cual indica que las variables si están correlacionadas y se puede aplicar PCA.

In [14]:
#@title Matriz de covarianza
matriz_cov = np.cov(df_data_std.T)
matriz_cov

array([[ 1.00075301,  0.91243298,  0.90354353,  0.91087956,  0.20849747,
         0.38976694,  0.40311684,  0.16612155,  0.79168451,  0.78039974,
         0.23633787,  0.57408716,  0.74762803,  0.71138844,  0.6296208 ,
         0.75762024,  0.40131247,  0.82705322],
       [ 0.91243298,  1.00075301,  0.99162006,  0.98041398,  0.26178409,
         0.34554933,  0.35546507,  0.15118559,  0.89692032,  0.88148878,
         0.25459516,  0.5768548 ,  0.69708465,  0.67951286,  0.55197618,
         0.67544505,  0.39078984,  0.85091251],
       [ 0.90354353,  0.99162006,  1.00075301,  0.98085506,  0.29733653,
         0.28817283,  0.29812184,  0.11958326,  0.84878505,  0.8412998 ,
         0.21958828,  0.5979175 ,  0.70572244,  0.69319964,  0.53266184,
         0.66295543,  0.40104558,  0.83497134],
       [ 0.91087956,  0.98041398,  0.98085506,  1.00075301,  0.13591733,
         0.38915282,  0.41236314,  0.19730809,  0.8268661 ,  0.80611318,
         0.26460448,  0.50592816,  0.64323551,  0.616

In [15]:
#@title Autovalores y autovectores
autovalores, autovector = np.linalg.eig(matriz_corr)
autovalores

array([9.17575683e+00, 3.89503599e+00, 1.21709332e+00, 8.45684220e-01,
       6.42252621e-01, 5.35558847e-01, 5.01499523e-01, 3.98050528e-01,
       2.61785818e-01, 2.32801087e-01, 1.10114420e-01, 1.00039556e-01,
       5.67059160e-02, 1.47281503e-02, 7.68631780e-03, 4.67018240e-03,
       1.68648038e-04, 3.68024308e-04])

In [16]:
#@title Porcentaje de varianza explicada por todas las componentes
pct_varianza_explicada = autovalores/sum(autovalores)
pct_varianza_explicada

array([5.09764268e-01, 2.16390888e-01, 6.76162957e-02, 4.69824567e-02,
       3.56807012e-02, 2.97532693e-02, 2.78610846e-02, 2.21139182e-02,
       1.45436565e-02, 1.29333937e-02, 6.11746775e-03, 5.55775309e-03,
       3.15032866e-03, 8.18230572e-04, 4.27017656e-04, 2.59454578e-04,
       9.36933546e-06, 2.04457949e-05])

In [17]:
#@title Selección del numero de componentes (Kaiser)
autovalores_kaiser = autovalores[autovalores >= 0.8]
autovalores_kaiser

array([9.17575683, 3.89503599, 1.21709332, 0.84568422])

In [18]:
#@title Porcentaje de varianza explicada de las componentes elegidas
varianza_explicada=[]
varianza_explicada_acum =[]

# Loop para capturar la varianza explicada y la varianza explicada acumulada
for i in autovalores_kaiser:
    varianza_explicada.append((i)/sum(autovalores)*100)
    varianza_explicada_acum.append(sum(varianza_explicada))


# Generar DataFrame
matriz_var_expl = pd.DataFrame({'pct_var_expl': varianza_explicada,
                                'pct_var_expl_acum': varianza_explicada_acum
                               },
                               index=np.linspace(1,len(autovalores_kaiser),len(autovalores_kaiser)))


# Visualizar matriz
matriz_var_expl

pct_var_expl  pct_var_expl_acum
1.0     50.976427          50.976427
2.0     21.639089          72.615516
3.0      6.761630          79.377145
4.0      4.698246          84.075391

Según los autovalores de Kaiser, deberíamos ocupar 4 componentes, ya que cumplen con el umbral de 0.8 y por lo tanto capturan una cantidad significativa de varianza. Con esta cantidad de componentes explicaríamos el 84.07% de la varianza. 

## Parte 3

In [ ]:
El LazyClassifier es una clase de clasificación diseñado para el análisis exploratorio de datos y para obtener de manera mas sencilla una 
idea sobre los modelos mas adecuados de clasificación para un conjunto de datos especifico sin entrenar todos los modelos manualmente. 

### Mejores algoritmos de clasificación

1. K-Nearest Neighbors (KNN): Algoritmo de que clasifica un punto en base a los k vecinos más cercanos en el espacio de características.
2. Regresión Logística: Algoritmo que utiliza una función logística para modelar la probabilidad de una clase.
3. Support Vector Machines (SVM): Encuentra el hiperplano que mejor separa las clases en el espacio de características.
4. Random Forest: Combina múltiples árboles de decisión para mejorar la precisión y reducir el overfitting.


## Parte 4

In [19]:
df = sns.load_dataset('iris')

In [20]:
df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [21]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB
None


In [22]:
print(df.describe())

       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


In [31]:
#@title Función para generar métricas
def fn_metricas_clasificacion(y_true, y_pred):
  print('Accuracy     : %.3f' % accuracy_score(y_true, y_pred))
  print("Precisión    : {}".format(precision_score(y_true, y_pred, average=None)))
  print("Recall       : {}".format(recall_score(y_true, y_pred, average=None)))
  print("Puntuación F1: {}".format(f1_score(y_true, y_pred, average=None)))
  print("Cohen Kappa  : {}".format(cohen_kappa_score(y_true, y_pred )))

In [32]:
#@title Crear función para visualizar métricas
def fn_metricas_info(modelo, y_verdadero, y_calculado):
    metrics = {
        'Accuracy': "%.5f" % accuracy_score(y_verdadero, y_calculado),
        'Precisión': "{}".format(precision_score(y_verdadero, y_calculado, average=None)),
        'Recall': "{}".format(recall_score(y_verdadero, y_calculado, average=None)),
        'Puntuación F1': "{}".format(f1_score(y_verdadero, y_calculado, average=None)),
        'Cohen Kappa': "{}".format(cohen_kappa_score(y_verdadero, y_calculado))
    }
    return pd.DataFrame(metrics, index=[modelo])

### Modelo Random Forest

In [33]:
#@title Variables del modelo y sus entrenamientos
X = df.drop('species', axis=1)
y = df['species']

# Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=2024)

In [34]:
#@title Instanciar algoritmo de clasificación RFC
rfc_clf = RandomForestClassifier()
# Aplicar modelo a datos de entrenamietno
classifier = rfc_clf.fit(X_train, y_train)

In [35]:
#@title Generar predicciones con datos test
y_pred_rfc = classifier.predict(X_test)

In [36]:
#@title Generar la matriz de confusión
confusion_matrix(y_test, y_pred_rfc)

array([[36,  0,  0],
       [ 0, 30,  6],
       [ 0,  1, 32]])

In [38]:
#@title Mostrar métricas
fn_metricas_clasificacion(y_test, y_pred_rfc)

Accuracy     : 0.933
Precisión    : [1.         0.96774194 0.84210526]
Recall       : [1.         0.83333333 0.96969697]
Puntuación F1: [1.         0.89552239 0.90140845]
Cohen Kappa  : 0.9001222992254383


In [39]:
#@title Mostrar Reporte
print(classification_report(y_test, y_pred_rfc))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        36
  versicolor       0.97      0.83      0.90        36
   virginica       0.84      0.97      0.90        33

    accuracy                           0.93       105
   macro avg       0.94      0.93      0.93       105
weighted avg       0.94      0.93      0.93       105



In [42]:
#@title Ajustes de Hiperparámetros
from sklearn.model_selection import GridSearchCV

# Generar las particiones
particiones = KFold(n_splits=10, shuffle=True, random_state=0)

# Definir los hiperparámetros y sus posibles valores
param_grid = {
    'n_estimators': [10,25,50],
    'max_depth' : [5,10,15],
    'criterion' : ['gini', 'entropy', 'log_loss'],
    'min_samples_split': [6,8,12],
    'min_samples_leaf': [4,8,12],
}

# Instanciar el modelo sin hiperparámetros
rf_clf = RandomForestClassifier(class_weight='balanced')

# Aplicar GridSearchCV  usando el diccionario de hiperparámetros
mejor_modelo_rfc = GridSearchCV(estimator=rf_clf, param_grid=param_grid, cv=particiones, scoring='accuracy')

# Entrenar el modelo con datos train
mejor_modelo_rfc.fit(X_train, y_train)

# Obtener los mejores parámetros
mejor_modelo_rfc.best_params_

{'criterion': 'gini',
 'max_depth': 5,
 'min_samples_leaf': 4,
 'min_samples_split': 6,
 'n_estimators': 50}

In [43]:
#@title Parametros estimados
mejor_modelo_rfc.best_estimator_

RandomForestClassifier(class_weight='balanced', max_depth=5, min_samples_leaf=4,
                       min_samples_split=6, n_estimators=50)

In [44]:
#@title Obtener el mejor modelo
modelo_final = mejor_modelo_rfc.best_estimator_
# Obtener predicciones con el set de test
y_pred = modelo_final.predict(X_test)
# Imprimir reporte de clasificación
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        36
  versicolor       1.00      0.86      0.93        36
   virginica       0.87      1.00      0.93        33

    accuracy                           0.95       105
   macro avg       0.96      0.95      0.95       105
weighted avg       0.96      0.95      0.95       105



In [45]:
#@title Métricas del modelo
metrica = fn_metricas_info('random_forest', y_test, y_pred)
metrica

Accuracy                           Precisión  \
random_forest  0.95238  [1.         1.         0.86842105]   

                                           Recall  \
random_forest  [1.         0.86111111 1.        ]   

                                    Puntuación F1         Cohen Kappa  
random_forest  [1.         0.92537313 0.92957746]  0.9286587851610273

In [46]:
df_resultados = metrica
df_resultados

Accuracy                           Precisión  \
random_forest  0.95238  [1.         1.         0.86842105]   

                                           Recall  \
random_forest  [1.         0.86111111 1.        ]   

                                    Puntuación F1         Cohen Kappa  
random_forest  [1.         0.92537313 0.92957746]  0.9286587851610273

El modelo de clasificación Random Forest, después de ajustar sus hiperparámetros, muestra un rendimiento sobresaliente con una alta exactitud (95.24%) y valores elevados en precisión, recall, F1 score, y Cohen's Kappa. Estas métricas indican que el modelo es muy efectivo en predecir correctamente las clases, capturando la mayoría de los verdaderos positivos y manteniendo un equilibrio saludable entre precisión y recall. En resumen, el modelo predice de manera muy precisa y confiable, siendo adecuado para su propósito de clasificación en este conjunto de datos.